src/c++/perf_analyzer/docs/examples/profile.py

```diff
@@ -175,73 +223,134 @@ def collect_periodic_latencies(args, filename):
-    return latencies
+        latencies.append(np.mean(bin) / 1_000_000)
+
+    profile_result.avg_periodic_t2t_latencies = latencies
 
 
 def collect_latencies(requests):
     # Example json demonstrating format:
     #   see client/src/c++/perf_analyzer/docs/examples/decoupled_output_file.json
     first_token_latencies = []
+    generation_latencies = []
     token_to_token_latencies = []
     requests = requests["experiments"][0]["requests"]
+    for r in requests:
+        init_request, responses = r["timestamp"], r["response_timestamps"]
+        first_token_latencies.append((responses[0] - init_request) / 1_000_000)
+        generation_latencies.append((responses[-1] - responses[0]) / 1_000_000)
+        token_to_token_latencies = []
+        for prev_res, res in pairwise(responses):
+            token_to_token_latencies.append((res - prev_res) / 1_000_000)
+    return first_token_latencies, generation_latencies, token_to_token_latencies
+
+
+def calculate_online_metrics(args, profile_result, filename):
+    """Calculate online metrics for more fine-grained performance information."""
+    if not args.stream:
+        return  # skip if offline
+
+    requests = load_json_data(filename)
+    latencies = collect_latencies(requests)
+    first_token_latencies, generation_latencies, token_to_token_latencies = latencies
+
+    profile_result.avg_first_token_latency = np.mean(first_token_latencies)
+    profile_result.avg_total_t2t_latency = np.mean(token_to_token_latencies)
+
+    profile_result.max_first_token_latency = max(first_token_latencies)
+    profile_result.min_first_token_latency = min(first_token_latencies)
+    profile_result.avg_first_token_latency = np.mean(first_token_latencies)
+    profile_result.p50_first_token_latency = np.percentile(
+        first_token_latencies, 50, method="lower"
+    )
+    profile_result.p90_first_token_latency = np.percentile(
+        first_token_latencies, 90, method="lower"
+    )
+    profile_result.p95_first_token_latency = np.percentile(
+        first_token_latencies, 95, method="lower"
+    )
+    profile_result.p99_first_token_latency = np.percentile(
+        first_token_latencies, 99, method="lower"
+    )
+
+    profile_result.max_gen_latency = max(generation_latencies)
+    profile_result.min_gen_latency = min(generation_latencies)
+    profile_result.avg_gen_latency = np.mean(generation_latencies)
+    profile_result.p50_gen_latency = np.percentile(
+        generation_latencies, 50, method="lower"
+    )
+    profile_result.p90_gen_latency = np.percentile(
+        generation_latencies, 90, method="lower"
+    )
+    profile_result.p95_gen_latency = np.percentile(
+        generation_latencies, 95, method="lower"
+    )
+    profile_result.p99_gen_latency = np.percentile(
+        generation_latencies, 99, method="lower"
+    )
+
+    token_latencies = [t / args.max_tokens for t in generation_latencies]
+    profile_result.avg_token_latency = np.mean(token_latencies)
```

Comment:
Can this have a more descriptive name? Maybe `avg_per_output_token_latency`? This is similar to the average per-output token latency from the vLLM benchmarking script, right?:

https://github.com/vllm-project/vllm/blob/main/benchmarks/benchmark_serving.py#L199-L202

Reply:
`avg_per_output_token_latency` sounds a bit weird to me. I will change to `avg_latency_per_output_token`.